### PR TITLE
feat(messaging): Phase 1 — PostgreSQL durable queue with LISTEN/NOTIFY worker

### PR DESCRIPTION
**Pull request by Sudarshan25092007**
## Summary

Part of #1731

This PR introduces the foundational, PostgreSQL-native messaging queue infrastructure for Phase 1.

### What this PR does

- Adds a `messaging` schema with `outbox`, `delivery_attempts`, and `dead_letter` tables
- Introduces an in-process worker using `LISTEN/NOTIFY` for low-latency job execution with a 5s polling fallback
- Implements atomic claim/lease locking using `FOR UPDATE SKIP LOCKED` and unique `claim_token` leases
- Adds stored procedure `messaging.reconcile_jobs()` for orphan recovery and retry scheduling
- Moves permanent failures to `messaging.dead_letter` when `retry_count` reaches `max_retries`
- Routes Phase 1 delivery through existing `EmailService.sendRaw()` path
- Single consolidated unit test suite (`backend/tests/unit/messaging.test.ts`)

> Phase 1 strictly validates core queueing infrastructure. It does not introduce SMS/Push providers, public SDKs, or admin UI endpoints.

---

### What this PR does not do (Deferred Scope)

| Area | Why it is out of scope | Planned phase |
|------|------------------------|---------------|
| Twilio / FCM providers | Validate queue behavior against existing email path first | Phase 3 (SMS), Phase 4 (Push) |
| `insforge.messaging.send()` SDK | Public SDK comes after queue core is stabilized | Phase 5 |
| Dead letter LIST / REVIVE APIs | Admin management UI endpoints | Phase 2 |
| User message status endpoint | Public status query API | Phase 2 |
| Webhook handlers | SendGrid / Mailgun webhook ingestion | Phase 3 |
| Rate-limiting UI / middleware | Internal queue validation doesn't need public rate limiting | Phase 2 |

---

### 1. `messaging.outbox` — Queue Table

| Column | Type | Purpose |
|--------|------|---------|
| `id` | `UUID PK` | Primary key (`gen_random_uuid()`) |
| `channel` | `TEXT CHECK ('email', 'sms', 'push')` | Channel identifier |
| `status` | `TEXT CHECK ('pending', 'claimed', 'sent', 'failed', 'dead')` | Outbox lifecycle status |
| `payload` | `JSONB` | Message payload (to, subject, body, metadata) |
| `idempotency_key` | `TEXT` | Optional key for deduplication |
| `claimed_by` | `TEXT` | Worker instance ID holding active claim |
| `claimed_at` | `TIMESTAMPTZ` | Timestamp when claimed |
| `lease_expires_at` | `TIMESTAMPTZ` | Deadline for active worker claim |
| `claim_token` | `UUID` | Unique token allocated per lease to prevent race conditions |
| `retry_count` | `INT` | Current retry attempt count |
| `max_retries` | `INT` | Max retry limit before DLQ promotion (default 5) |
| `next_attempt_at` | `TIMESTAMPTZ` | Scheduled retry timestamp |
| `provider_message_id` | `TEXT` | Returned message ID from delivery provider |
| `error_message` | `TEXT` | Last failure error details |
| `created_at`, `updated_at` | `TIMESTAMPTZ` | Record timestamps |

### 2. `messaging.delivery_attempts` — Audit Log

| Column | Type | Purpose |
|--------|------|---------|
| `id` | `UUID PK` | Audit row ID |
| `message_id` | `UUID` | Reference to outbox message |
| `attempt_number` | `INT` | Sequential attempt index |
| `worker_id` | `TEXT` | Worker instance ID |
| `status` | `TEXT` | Outcome status (`sent`, `failed`) |
| `error_message` | `TEXT` | Error details if failed |
| `duration_ms` | `INT` | Attempt execution time |
| `attempted_at` | `TIMESTAMPTZ` | Timestamp of attempt |

### 3. `messaging.dead_letter` — Permanent Failures

| Column | Type | Purpose |
|--------|------|---------|
| `id` | `UUID PK` | Original outbox message ID |
| `channel`, `payload`, `idempotency_key` | Copied columns | Message snapshot |
| `retry_count`, `max_retries` | `INT` | Final retry counters |
| `error_message` | `TEXT` | Cause of permanent failure |
| `created_at`, `moved_at` | `TIMESTAMPTZ` | Timestamps when created and moved to DLQ |

---

## Worker Lifecycle & Flow

1. **Enqueue**: Request hits `POST /api/messaging/send` → validated & inserted into `messaging.outbox` with `status = 'pending'`.
2. **Notification**: Postgres trigger `trg_messaging_notify_new_job` emits `NOTIFY 'messaging_new_job'`.
3. **Claim**: `MessagingWorker` wakes up and executes:
   ```sql
   UPDATE messaging.outbox
   SET status = 'claimed', claimed_by = $1, lease_expires_at = NOW() + INTERVAL '60s', claim_token = gen_random_uuid()
   WHERE id = (
     SELECT id FROM messaging.outbox WHERE status = 'pending' AND next_attempt_at <= NOW()
     ORDER BY next_attempt_at ASC, created_at ASC LIMIT 1 FOR UPDATE SKIP LOCKED
   ) RETURNING *;
4. Dispatch: Worker invokes EmailService.sendRaw(payload, abortSignal).
5. Mark Sent: Worker calls markSent() validating both claimed_by and claim_token.
6. Failure / Retry: If dispatch fails, markFailed() increments retry_count and sets next_attempt_at using exponential backoff. If retry_count + 1 >= max_retries, it atomically moves the record to messaging.dead_letter and deletes it from messaging.outbox.

## Reconciliation Stored Procedure
messaging.reconcile_jobs(p_backoff_base_seconds INT, p_jitter_pct FLOAT) recovers orphaned leases:

- Scans messaging.outbox where status = 'claimed' and lease_expires_at <= NOW().
- If retry_count + 1 >= max_retries, moves record to messaging.dead_letter.
- Otherwise resets status = 'pending', increments retry_count, and schedules next_attempt_at with exponential backoff and jitter.

---

### Known Limitations (Phase 1)

- **At-least-once delivery only.** During in-flight lease loss, duplicate emails are possible despite deterministic Message-ID and pre-send audit checks.
- **SMTP acceptance and PostgreSQL persistence are not atomic.** Exact-once guarantee requires provider webhook confirmation (Phase 3) and is not claimed in this PR.
- **Single-replica worker.** The in-process worker runs inside the existing Express container. Multi-replica deployments share queue state via Postgres, but local fallback markers are not visible across replicas.


<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Replaces in-request email sends with a PostgreSQL-backed durable outbox and a background worker. Requests now enqueue and return 202; a `LISTEN/NOTIFY` worker delivers with retries and a dead-letter queue for at-least-once delivery (duplicates possible; mitigated by pre-send audit and deterministic Message-ID).

- Database/migration: new `messaging` schema (`outbox`, `delivery_attempts`, `dead_letter`), insert `NOTIFY` trigger, and `messaging.reconcile_jobs()` for lease recovery. Action: run `065_create-messaging-schema.sql`.
- API/behavior: admin-only POST `/api/messaging/send` validates, enqueues, and returns `{messageId, status}`; non-email channels return 501; idempotency conflicts return 409 on payload mismatch; if the key is in DLQ, returns `dead` with `isDuplicate`.
- Worker/runtime: background worker starts on server boot; dedicated `LISTEN` client with 5s polling fallback and auto-reconnect; startup race fix (assign listen client before connect, recheck `isRunning`); `FOR UPDATE SKIP LOCKED` claims with per-lease `claim_token`; periodic lease renewal; throttled reconciliation; max concurrency; graceful shutdown; durable attempt logging before releasing the claim; pre-send audit skips if a prior “sent” exists.
- Providers: `AbortSignal` pre-flight added; SMTP sets deterministic RFC 5322 Message-ID from SHA-256 of the idempotency key (or message id); cloud provider forwards the signal. In-flight SMTP sends cannot be canceled.
- Rollout/config: apply the migration and set env before deploy. Required: `MESSAGING_LEASE_DURATION_SECONDS`, `MESSAGING_MAX_RETRY_ATTEMPTS`, `MESSAGING_BACKOFF_BASE_SECONDS`, `MESSAGING_JITTER_PERCENT`; optional: `MESSAGING_MAX_CONCURRENCY`, `MESSAGING_RECONCILIATION_INTERVAL_SECONDS`.

<sup>Written for commit c6d1705571983151a77e894d8c6469dddc90a3cb. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1795?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg"><img alt="Review in cubic" src="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"></picture></a>

<!-- End of auto-generated description by cubic. -->







<!-- Macroscope's pull request summary starts here -->
<!-- Macroscope will only edit the content between these invisible markers, and the markers themselves will not be visible in the GitHub rendered markdown. -->
<!-- If you delete either of the start / end markers from your PR's description, Macroscope will append its summary at the bottom of the description. -->
> [!NOTE]
> ### Add PostgreSQL durable outbox queue with LISTEN/NOTIFY background worker for messaging
> - Adds a new `messaging` schema via [065_create-messaging-schema.sql](https://github.com/InsForge/InsForge/pull/1795/files#diff-4d4359576d0738a4b87e6a99ec9307deccb01b87fe05734bd94b7c277dd6bacd) with `outbox`, `delivery_attempts`, and `dead_letter` tables, plus a PostgreSQL trigger that fires `NOTIFY messaging_new_job` on insert.
> - Introduces [MessagingQueueService](https://github.com/InsForge/InsForge/pull/1795/files#diff-07915030289c4908ae132df4d4a48666b084e9d3f0a6b6a34f7ef76c58829d7d) implementing idempotent enqueue, atomic claim/lease via `SKIP LOCKED`, sent/failed state transitions, exponential backoff with jitter, and dead-letter promotion.
> - Adds [MessagingWorker](https://github.com/InsForge/InsForge/pull/1795/files#diff-ad759ffb3291b92a67bc238ea9cd4988d4ebf32e743f993651fb96e2161b4c68) as a background process that listens for notifications, polls as fallback, enforces concurrency limits, renews leases via heartbeat, and aborts in-flight sends on lease loss.
> - Exposes `POST /api/messaging/send` (admin-only) in [index.routes.ts](https://github.com/InsForge/InsForge/pull/1795/files#diff-8102645112bd707bcd458e73dbff68596605d439d735f6c298c3cc0f7e7500e5) returning 202 with `messageId` and status.
> - Email providers now accept `AbortSignal` and an `idempotencyKey` that maps to a deterministic RFC 5322 `Message-ID`.
> - Risk: the worker starts at server boot and requires a dedicated long-lived `pg` client for `LISTEN`; a failure to connect on startup logs an error and falls back to polling only.
>
> <!-- Macroscope's review summary starts here -->
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 401969f.</sup>
> <!-- Macroscope's review summary ends here -->
>
<!-- Macroscope's pull request summary ends here -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **New Features**
  * Added an admin-protected API for submitting email messages.
  * Added reliable queuing with duplicate prevention, retries, recovery, dead-letter handling, and delivery tracking.
  * Added background processing with configurable concurrency and reconciliation.
  * Added cancellation and idempotency support for email delivery.
* **Configuration**
  * Added settings for concurrency, leases, retries, backoff, and jitter.
* **Bug Fixes**
  * Improved handling of interrupted deliveries and transient failures.
* **Tests**
  * Added comprehensive messaging and delivery coverage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by Sudarshan25092007** [comment]
@Fermionic-Lyu - opened the draft POC PR for Phase 1 as discussed. 

Additions across the migration, queue service, worker, dead-letter handling, and admin routes. The worker is in-process (consistent with current `concurrently` setup) and routes Phase 1 through `EmailService.sendRaw()` only.

Since it's a draft, I kept it open for architectural feedback before I fill in the remaining edge cases and tests. Let me know if the overall structure matches what you had in mind, or if I should pivot on anything before polishing.
And also can you manually remove the label of `needs-issue` from the bot.

No rush - whenever you have bandwidth.

**Comment by Sudarshan25092007** [comment]
Just a small ping on this @Fermionic-Lyu , as this is draft it might have lost in queue. Once you review this, we are good to move with the next implementation.

**Comment by Sudarshan25092007** [comment]
Hey @Fermionic-Lyu  , Resolved the merge conflicts and moved the migration from 60 to 62.
 Will iterate it further after your reviews on this.

**Comment by Sudarshan25092007** [comment]
@cubic-dev-ai review

File: backend/src/infra/database/migrations/063_create-messaging-schema.sql [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+CREATE TRIGGER trg_messaging_new_job
+  AFTER INSERT ON messaging.outbox
+  FOR EACH ROW
+  EXECUTE FUNCTION messaging.notify_on_job_insert();
+
+-- ============================================================================
+-- RECONCILIATION FUNCTION
+-- ============================================================================
+
+CREATE OR REPLACE FUNCTION messaging.reconcile_jobs(
```
[the remark this answers is not in the document]
Comment (Sudarshan25092007) [reply backend/src/infra/database/migrations/063_create-messaging-schema.sql:90]: A PostgreSQL integration test for reconcile_jobs() requires a live DB fixture -  this is planned for Phase 2 alongside the heartbeat table and monitoring work.

**Comment by Sudarshan25092007** [comment]
@cubic-dev-ai review

**Comment by Sudarshan25092007** [comment]
@cubic-dev-ai review

**Comment by Sudarshan25092007** [comment]
@cubic-dev-ai review

**Comment by Sudarshan25092007** [comment]
@coderabbitai review

**Comment by Sudarshan25092007** [comment]
@greptileai review

**Comment by Sudarshan25092007** [comment]
@coderabbitai resume

**Comment by Sudarshan25092007** [comment]
@coderabbitai review

**Comment by Sudarshan25092007** [comment]
@Fermionic-Lyu, I've iterated through the automated reviews and stabilized the core POC structure:

- Postgres-native outbox with LISTEN/NOTIFY worker
- Atomic claim/lease via FOR UPDATE SKIP LOCKED + claim_token
- Exponential backoff with jitter, dead-letter promotion at max retries
- Phase 1 routes through EmailService.sendRaw()

Two edge cases remain where I'd value your architectural guidance rather than keep guessing:

1. **Reconciliation scheduling:** `messaging.reconcile_jobs()` exists in SQL but the worker never explicitly invokes it. Should the worker run this on a timer, or should pg_cron handle scheduling entirely? Currently expired claims can sit unrecovered if LISTEN/NOTIFY is the only driver.

2. **In-flight lease loss during SMTP:** If a worker's lease expires while Nodemailer is mid-send, the AbortSignal pattern doesn't actually cancel the SMTP transaction. The stale send completes while reconciliation requeues the message, risking duplicates. Is this acceptable for Phase 1, or do you want a different guard (e.g., lease renewal during long sends)?

(I was just over-engineering all the issues that were pointed out by bots then i came to know that this should only be POC, small and concise for your comments on this! Cut down the lines of code written and kept the initial stuff infront you for review)
Happy to fix whichever direction you prefer and work under your direction. Just don't want to over-engineer the POC in a vacuum.

File: backend/src/providers/email/smtp.provider.ts [hunk 2]
```diff
@@ -123,8 +123,22 @@ export class SmtpEmailProvider implements EmailProvider {
...
     const config = await this.getRequiredConfig();
+    if (signal?.aborted) {
+      throw new AppError(
+        'Email sending aborted due to lost claim',
+        500,
+        ERROR_CODES.EMAIL_SMTP_SEND_FAILED
+      );
+    }
 
     await this.send(
```
Comment (Fermionic-Lyu) [inline backend/src/providers/email/smtp.provider.ts:236]: The AbortSignal is never passed into `this.send()` or `transporter.sendMail()`. Once this call starts, losing the lease cannot stop the SMTP transaction, so reconciliation can requeue the job while the stale worker still delivers it and produce duplicate emails. Please make the in-flight SMTP operation lease-aware and use a stable provider idempotency mechanism, since closing a transport cannot revoke a message the SMTP server already accepted.

File: backend/src/services/messaging/worker.service.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    try {
+      await this.queueService.reconcile().catch(() => {});
+      const message = await this.queueService.claim(this.workerId);
+      if (!message) {
+        return;
+      }
+
+      this.activeProcessingCount++;
```
Comment (Fermionic-Lyu) [inline backend/src/services/messaging/worker.service.ts]: `activeProcessingCount` is incremented only after both `reconcile()` and `claim()` have awaited. A burst of notification callbacks can all pass the earlier limit check and then increment, so `maxConcurrency` is not actually enforced. Please acquire a semaphore slot before the first await, release it in `finally`, and read the limit from `appConfig.messaging.maxConcurrency` instead of hardcoding 5.

File: backend/src/services/messaging/queue.service.ts [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+            );
+          }
+          await client.query('COMMIT');
+          return { id: dbRow.id, status: dbRow.status as MessageStatus };
+        }
+      }
+
+      const maxRetries = appConfig.messaging.maxRetryAttempts;
+      const sql = `
+        INSERT INTO messaging.outbox (channel, status, payload, idempotency_key, retry_count, max_retries, next_attempt_at)
```
Comment (Fermionic-Lyu) [inline backend/src/services/messaging/queue.service.ts]: `SELECT ... FOR UPDATE` only locks an existing row. Two first-time requests with the same idempotency key can both see no row and race on this insert; the loser receives an unhandled `23505`/500 instead of the original result. Please make creation atomic with `INSERT ... ON CONFLICT` or a key-scoped advisory lock, then load and compare the winning record.

File: backend/src/infra/database/migrations/063_create-messaging-schema.sql [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                id, channel, payload, idempotency_key, retry_count, max_retries, error_message, created_at, moved_at
+            )
+            SELECT id, channel, payload, idempotency_key, retry_count + 1, max_retries, 'Lease expired (max retries reached)', created_at, NOW()
+            FROM messaging.outbox WHERE id = r.id;
+
+            DELETE FROM messaging.outbox WHERE id = r.id;
+        ELSE
+            -- TODO: Use configurable backoff_base_seconds parameter instead of hardcoded 5s base
+            v_jitter_factor := 1.0 + ((random() * 2.0 - 1.0) * p_jitter_pct);
+            v_delay_seconds := (5.0 * power(2, r.retry_count)) * v_jitter_factor;
```
Comment (Fermionic-Lyu) [inline backend/src/infra/database/migrations/063_create-messaging-schema.sql]: `p_backoff_base_seconds` is ignored here, and the worker invokes reconciliation with defaults on every poll/notification, so the configured backoff, jitter, and reconciliation interval are ineffective. Please use the function parameter and schedule reconciliation from `appConfig.messaging` at the configured interval.

**Review by Fermionic-Lyu — CHANGES_REQUESTED** [review]
Blocking delivery, concurrency, idempotency, and reconciliation issues remain. Please address the inline comments before merging this queue infrastructure.

**Comment by Sudarshan25092007** [comment]
@coderabbitai review
@greptileai review

**Comment by Sudarshan25092007** [comment]
@coderabbitai review

**Comment by Sudarshan25092007** [comment]
@coderabbitai review
@cubic-dev-ai review
@greptileai review

**Comment by Sudarshan25092007** [comment]
@coderabbitai resume
@coderabbitai review

File: backend/src/services/messaging/worker.service.ts [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+            if (!renewed) {
+              logger.warn(`Failed to renew lease for message ${message.id}. Aborting process.`);
+              abortController.abort();
+              clearInterval(intervalId);
+            }
+          })
+          .catch((err) => {
+            logger.error(`Error renewing lease for message ${message.id}`, { error: err });
+            abortController.abort();
+            clearInterval(intervalId);
```
[the remark this answers is not in the document]
Comment (Sudarshan25092007) [reply backend/src/services/messaging/worker.service.ts:286]: Acknowledged. Conservative abort on any renewal failure (including transient DB errors) is intentional for Phase 1 to prevent stale-worker duplicates. Documented under Known Limitations: at-least-once delivery. Exact-once requires provider webhook confirmation (Phase 3).

**Comment by Sudarshan25092007** [comment]
@Fermionic-Lyu, squashed to 6 commits, all inline feedback addressed:

**Your blockers:** AbortSignal pre-flight only | maxConcurrency sync guard | atomic idempotency CTE | reconcile uses config param.

**Additional hardening**: advisory-locked audit insert | reconnect timer cleanup | LISTEN shutdown race (client assigned early + isRunning recheck + UNLISTEN removed with timeout) | dead-letter ON CONFLICT (id).

> Greptile's remaining duplicate-send concern is the **documented** Phase 1 at-least-once boundary (see PR description → Known Limitations). Exact-once requires Phase 3 webhooks.

Ready for your re-review. No rush.
